### DISCUSSION TITLE
Question: Use of autoForward in conjunction with onDone and onError

### DISCUSSION THREAD

**Discussion by kingcars** [statelyai/xstate#1980] [Q&A]
## Bug or feature request?

Potentially a bug, but could just be incorrect implementation patterns

## Description:

I'm currently working on a complex admin application which requires heavy use of hierarchical machines and parallel states; essentially, we have a main `parent` machine for each page in the admin, and each modal/banner/etc within that page gets spawned as a child machine or parallel state. The modals and banners talk to their respective machines by sending events through the parent machine. We've been using `autoForward` to reach these invoked machines in order to cut down on the high amount of `send` actions that would be required otherwise. However, in doing so, we've stumbled across some behavior that is either a bug or simply an issue with this pattern which we've been implementing. If it's the latter, then I'd very much be interested in discussing the recommended patterns for the examples below.

Issue 1: When a child machine is invoked and reaches its `final` state, the `onDone` transition will throw a warning in console:

`Warning: Event "done.invoke.childMachineInvoke" was sent to stopped service "childMachine". This service has already reached its final state, and will not transition.
Event: {"type":"done.invoke.childMachineInvoke"}`

Note: I noticed in testing that this warning does not pop up if the child machine reaches its `final` state via a `delay` as opposed to an explicit event call.

Issue 2: When a `Promise` in invoked within a state and hits `reject`, the resulting `onError` event will break with an error in the console:

`Missing onError handler for invocation 'failingApiCall', error was 'Error: Api call failed!'. Stacktrace was 'Error: Api call failed!`

Note: This issue can be worked around by setting up the call within its own machine as opposed to a state.

## Link to reproduction or proof-of-concept:

https://codesandbox.io/s/hopeful-sinoussi-k75h7

**Comment by Andarist** [comment]
Thanks for the great repro case! Things like this shortens the feedback loop for reporters - they allow us, maintainers, to jump into answering **much** quicker (or at all, as we just can't always afford to setup repros on our own and issues get lost).

### issue number 2 (the third test in your repro)
it's not the parent machine that fails, but it's rather the inner machine (the one that has `autoForward` set on it). As this error event is an external event it is rightfully forwarded to all actors with `autoForward` flag as per [mainEventLoop](https://www.w3.org/TR/scxml/#mainEventLoop) algorithm. 

We might want to introduce better event filtering for `onError` in the future, but this will have to be yet discussed and implemented - most likely in the v5 which we have already started working on... unless you would like to explore implementing a fix for this one after consulting the appropriate solution with us for this problem. I'm afraid though that the final solution is not that trivial due to how this is currently implemented (and given that those errors are rethrown from [here](https://github.com/davidkpiano/xstate/blob/f972d0054a6449c8efd50919cdbc72fb4ebdcc12/packages/core/src/interpreter.ts#L720) and given that JS stack trace is shared during this nested, synchronous, forwarding). It's certainly not impossible, but would have to be done with care.

### issue number 1 (the forth test in your repro)

I think this happens because the child machine reaches its final state DURING forwarding (eh, those  pesky synchronous edge cases 😠). I will yet have to confirm this in full, but if that's the case then the fix will be quite straightforward.

**Comment by kingcars** [comment]
Thanks for the quick response! I had a suspicion that `autoForward` in general is a really difficult and tedious feature to implement with all the potential for corner cases, so I definitely understand. For issue 2, we're currently implementing the workaround from the demo page of simply moving our API calls off of the parent machines and into child machines, which isn't a big deal and probably cleaner in the long run. Let me know if you need anything else.

**Comment by Andarist** [comment]
I've diagnosed the issue number 1 further - it's actually simpler than I have thought. 

`done.invoke.id` events are just regular events in SCXML, so when the invoking sessions receives such it process it "as normal" and thus it also forwards it to the session which has generated it. SCXML doesn't describe how to manage your invoked sessions at all, so there is no guidance on that.

I think the best that we could do is to implement removing the stopped child from the children collection, but currently, it seems that this might be harder than I would have thought. The problem is that we don't have access to our own children's statuses, because they are just Actors and those don't expose such information. There also might be children with the same ID in different interpreters so we cant just remove a particular child upon receiving `done.invoke.id` (because events might be forwarded etc).

My best bet right now is to introduce a special global semaphore-like flag that would be able to silent those warnings, because we still want to have this warning when a user sends an event directly to a stopped machine, but at the same time we dont want to show it if it happens during our internal processing. @davidkpiano WDYT?

**Comment by semopz** [comment]
> I think the best that we could do is to implement removing the stopped child from the children collection, but currently, it seems that this might be harder than I would have thought. The problem is that we don't have access to our own children's statuses, because they are just Actors and those don't expose such information. There also might be children with the same ID in different interpreters so we cant just remove a particular child upon receiving `done.invoke.id` (because events might be forwarded etc).

I have a somewhat related problem. I have a service that spawns actors which eventually go into a 'final' state. I have a cleanup function in my `assign` callback to get rid of the stopped actors from the context but I've subsequently found they remain as 'children' so the warning messages ramp up due to autoForward hitting a growing number of stale actors. AFAIK, there's no way to set up a way to auto-delete those children as part of the machine config. I could grab the service from `useService` and do the cleanup inside a React component but that feels wrong (possibly cause other issues?).

Any advise on how to handle this properly for now?

**Comment by Andarist** [comment]
You could experiment with `actions.pure(ctx => actions.stop(ctx.actorRef.id))`. This is just a wild guess of mine, seems like we need to improve the story around this use case.

**Comment by semopz** [comment]
This is another thing I'm not too sure about: isn't it redundant to stop a service that is already in its final state? Is your line of thinking that `actions.stop` should remove the actor from the *parent's* children?

As for pure, I'm not too sure how to use it https://github.com/davidkpiano/xstate/issues/1014

**Comment by Andarist** [comment]
Well, we probably should clear children automatically upon receiving `done.invoke.ID` event. I've proposed this purely as a temporary workaround.

I've looked into the problem with `pure` and responded back in that thread.

**Comment by davidkpiano** [comment]
> My best bet right now is to introduce a special global semaphore-like flag that would be able to silent those warnings, because we still want to have this warning when a user sends an event directly to a stopped machine, but at the same time we dont want to show it if it happens during our internal processing. @davidkpiano WDYT?

I'll look into why `autoForward` is still forwarding to stopped machines. However, we shouldn't have any special-casing or filtering for autoForward - if the developer decides to use `autoForward`, they should be aware that _all_ events will be sent to the relevant actor.

**Comment by semopz** [comment]
Let me know if this is unrelated and I'll log a separate issue.
@Andarist I appreciate you were proposing a workaround but I was just wandering what `actions.stop` does for a services in its final state.

@davidkpiano I agree. Question is, how do I (as a developer) remove a child that I no longer have a need for (so that it isn't autoForwarded any further events)?

Have a look at this (instructions included). The point I'm trying to make is, there doesn't seem to be a way to remove stale actors:
https://codesandbox.io/s/stopping-stopped-actors-nj1xb

**Comment by Andarist** [comment]
Oh, this repro explains to me your problem - you should not call imperatively things on your actors. Use events! It's hard to quickly adjust your demo because of the `autoForward` in use, but if you want to stop a child from the outside you should send an event to it so it can stop on its own.

Apart from that, I've noticed why we don't clear things correctly and gonna prepare a PR for it in a moment.

**Comment by semopz** [comment]
Ok I think we're on the same page. In real life, my actors are sent events and the actors themselves decide whether to go into a 'final' state or not. My confusion up to this point was this: when an actor is in a final state, is it up to the developer to clean things up?

It seems at this point, the answer is NO and that `xstate` should do the cleanup (when it comes to children). As for actor refs in the `context`, that is up to the developer. I'm not sure what V5 will look like but I think that a lot of confusion can be spared if the docs explain the full lifecyle of a parent-child setup under different scenarios.

**Comment by Andarist** [comment]
> It seems at this point, the answer is NO and that xstate should do the cleanup (when it comes to children). As for actor refs in the context, that is up to the developer.

Exactly :+1:
